src/sql-parser/src/ast/defs/name.rs

```diff
@@ -29,12 +30,202 @@ use crate::ast::{AstInfo, QualifiedReplica};
+
+        Ident(s)
+    }
+
+    /// Generate a valid [`Ident`] with the provided `prefix` and `suffix`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use mz_sql_parser::ast::{Ident, IdentError};
+    ///
+    /// let good_id =
+    ///   Ident::try_generate_name("hello", "_world", |_| Ok::<_, IdentError>(true)).unwrap();
+    /// assert_eq!(good_id.as_str(), "hello_world");
+    ///
+    /// // Return invalid once.
+    /// let mut attempts = 0;
+    /// let one_failure = Ident::try_generate_name("hello", "_world", |_candidate| {
+    ///     if attempts == 0 {
+    ///         attempts += 1;
+    ///         Ok::<_, IdentError>(false)
+    ///     } else {
+    ///         Ok(true)
+    ///     }
+    /// })
+    /// .unwrap();
+    ///
+    /// // We "hello_world" was invalid, so we appended "_1".
+    /// assert_eq!(one_failure.as_str(), "hello_world_1");
+    /// ```
+    pub fn try_generate_name<P, S, F, E>(prefix: P, suffix: S, mut is_valid: F) -> Result<Self, E>
+    where
+        P: Into<String>,
+        S: Into<String>,
+        E: From<IdentError>,
+        F: FnMut(&Ident) -> Result<bool, E>,
+    {
+        const MAX_ATTEMPTS: usize = 1000;
+
+        let prefix: String = prefix.into();
+        let suffix: String = suffix.into();
+
+        // First just append the prefix and suffix.
+        let mut candidate = Ident(prefix.clone());
+        candidate.append_lossy(suffix.clone());
+        if is_valid(&candidate)? {
+            return Ok(candidate);
+        }
+
+        // Otherwise, append a number to the back.
+        for i in 1..MAX_ATTEMPTS {
+            let mut candidate = Ident(prefix.clone());
+            candidate.append_lossy(format!("{suffix}_{i}"));
+
+            if is_valid(&candidate)? {
+                return Ok(candidate);
+            }
+        }
+
+        // Couldn't find any valid name!
+        Err(E::from(IdentError::FailedToGenerate {
+            prefix,
+            suffix,
+            attempts: MAX_ATTEMPTS,
+        }))
+    }
+
+    /// Append the provided `suffix`, truncating `self` as necessary to satisfy our invariants.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use mz_sql_parser::{
+    ///     ident,
+    ///     ast::Ident,
+    /// };
+    ///
+    /// let mut id = ident!("🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵🔵");
+    /// id.append_lossy("🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴");
+    ///
+    /// // We truncated the original ident, removing all '🔵' chars.
+    /// assert_eq!(id.as_str(), "🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🟢🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴");
+    /// ```
+    pub fn append_lossy<S: Into<String>>(&mut self, suffix: S) {
+        // Make sure our suffix at least leaves a bit of room for the original ident.
+        const MAX_SUFFIX_LENGTH: usize = Ident::MAX_LENGTH - 8;
+
+        let mut suffix: String = suffix.into();
+        mz_ore::soft_assert!(suffix.len() <= MAX_SUFFIX_LENGTH);
```

Comment:
You may want to add this requirement to the doc-comments.

Reply:
Added!